File: src/ExpoJpushModule.ts
import { NativeModule, requireNativeModule } from "expo";

import { ExpoJpushModuleEvents } from "./ExpoJpush.types";

declare class ExpoJpushModule extends NativeModule<ExpoJpushModuleEvents> {
  /** 初始化 JPush 并返回当前设备的 registrationId */
  init(options: { debug: boolean }): Promise<string>;

  /** 获取当前设备的 registrationId */
  getRegistrationID(): Promise<string>;

  /** 设置角标数量 */
  setBadgeNumber(params: { badge: number }): Promise<void>;

  // ---- Tag 操作 ----

  /** 覆盖设置 tags（会替换现有全部 tags） */
  setTags(params: { tags: string[]; seq: number }): Promise<void>;

  /** 新增 tags（追加到现有 tags） */
  addTags(params: { tags: string[]; seq: number }): Promise<void>;

  /** 删除指定 tags */
  deleteTags(params: { tags: string[]; seq: number }): Promise<void>;

  /** 清空所有 tags */
  cleanTags(params: { seq: number }): Promise<void>;

  /** 查询全部 tags */
  getAllTags(params: { seq: number }): Promise<void>;

  // ---- Alias 操作 ----

  /** 设置 alias */
  setAlias(params: { alias: string; seq: number }): Promise<void>;

  /** 删除 alias */
  deleteAlias(params: { seq: number }): Promise<void>;

  /** 查询当前 alias */
  getAlias(params: { seq: number }): Promise<void>;

  // ---- 手机号码 ----

  /** 设置手机号码（用于"推送不到短信到"功能） */
  setMobileNumber(params: { mobileNumber: string }): Promise<void>;

  // ---- 应用内消息页面追踪 ----

  /** 进入页面（配合应用内消息使用，需与 pageLeave 配套调用） */
  pageEnterTo(params: { pageName: string }): Promise<void>;

  /** 离开页面（配合应用内消息使用，需与 pageEnterTo 配套调用） */
  pageLeave(params: { pageName: string }): Promise<void>;

  // ---- 本地通知 ----

  /** 添加本地通知 */
  addLocalNotification(params: {
    /** 通知 ID（整数） */
    id: number;
    title: string;
    content: string;
    /** 触发延迟（毫秒），相对当前时间。例如 3000 = 3秒后触发 */
    fireTime: number;
    extras?: Record<string, unknown>;
    /** Android 自定义通知布局文件名 */
    builderName?: string;
    /** 通知分类（Android: category；iOS: categoryIdentifier） */
    category?: string;
    /** Android 通知优先级 */
    priority?: number;
  }): Promise<void>;

  /** 移除指定本地通知 */
  removeLocalNotification(params: { id: string }): Promise<void>;

  /** 清除所有本地通知 */
  clearLocalNotifications(): Promise<void>;
}

export default requireNativeModule<ExpoJpushModule>("ExpoJpush");

File: ios/ExpoJpushModule.swift
import ExpoModulesCore

public class ExpoJpushModule: Module {
  private let logPrefix = "[expo-jpush][ios][module]"

  public func definition() -> ModuleDefinition {
    Name("ExpoJpush")

    Events(ExpoJpushEvent.all)

    OnCreate {
      ExpoJpushBridge.shared.attach(module: self)
    }

    OnDestroy {
      ExpoJpushBridge.shared.detach(module: self)
    }

    // ---- 基础 ----

    AsyncFunction("init") { (options: [String: Any]?) -> String in
      ExpoJpushBridge.shared.setupIfNeeded(options: options)
      let registrationId = ExpoJpushBridge.shared.registrationID()
      if !registrationId.isEmpty {
        ExpoJpushBridge.shared.emit(
          name: ExpoJpushEvent.registration,
          payload: ["registrationId": registrationId]
        )
      }
      return registrationId
    }

    AsyncFunction("getRegistrationID") { () -> String in
      return ExpoJpushBridge.shared.registrationID()
    }

    AsyncFunction("setBadgeNumber") { (params: [String: Any]?) in
      let badgeValue = (params?["badge"] as? NSNumber)?.intValue ?? 0
      ExpoJpushNativeBridge.setBadge(Int(badgeValue))
    }

    // ---- Tag 操作 ----

    AsyncFunction("setTags") { (params: [String: Any]) in
      guard let tags = params["tags"] as? [String] else { return }
      let seq = (params["seq"] as? NSNumber)?.intValue ?? 0
      ExpoJpushBridge.shared.setTags(tags, seq: seq)
    }

    AsyncFunction("addTags") { (params: [String: Any]) in
      guard let tags = params["tags"] as? [String] else { return }
      let seq = (params["seq"] as? NSNumber)?.intValue ?? 0
      ExpoJpushBridge.shared.addTags(tags, seq: seq)
    }

    AsyncFunction("deleteTags") { (params: [String: Any]) in
      guard let tags = params["tags"] as? [String] else { return }
      let seq = (params["seq"] as? NSNumber)?.intValue ?? 0
      ExpoJpushBridge.shared.deleteTags(tags, seq: seq)
    }

    AsyncFunction("cleanTags") { (params: [String: Any]) in
      let seq = (params["seq"] as? NSNumber)?.intValue ?? 0
      ExpoJpushBridge.shared.cleanTags(seq: seq)
    }

    AsyncFunction("getAllTags") { (params: [String: Any]) in
      let seq = (params["seq"] as? NSNumber)?.intValue ?? 0
      ExpoJpushBridge.shared.getAllTags(seq: seq)
    }

    // ---- Alias 操作 ----

    AsyncFunction("setAlias") { (params: [String: Any]) in
      guard let alias = params["alias"] as? String else { return }
      let seq = (params["seq"] as? NSNumber)?.intValue ?? 0
      ExpoJpushBridge.shared.setAlias(alias, seq: seq)
    }

    AsyncFunction("deleteAlias") { (params: [String: Any]) in
      let seq = (params["seq"] as? NSNumber)?.intValue ?? 0
      ExpoJpushBridge.shared.deleteAlias(seq: seq)
    }

    AsyncFunction("getAlias") { (params: [String: Any]) in
      let seq = (params["seq"] as? NSNumber)?.intValue ?? 0
      ExpoJpushBridge.shared.getAlias(seq: seq)
    }

    // ---- 手机号码 ----

    AsyncFunction("setMobileNumber") { (params: [String: Any]) in
      guard let mobileNumber = params["mobileNumber"] as? String else { return }
      ExpoJpushBridge.shared.setMobileNumber(mobileNumber)
    }

    // ---- 应用内消息页面追踪 ----

    AsyncFunction("pageEnterTo") { (params: [String: Any]) in
      guard let pageName = params["pageName"] as? String else { return }
      ExpoJpushBridge.shared.pageEnterTo(pageName)
    }

    AsyncFunction("pageLeave") { (params: [String: Any]) in
      guard let pageName = params["pageName"] as? String else { return }
      ExpoJpushBridge.shared.pageLeave(pageName)
    }

    // ---- 本地通知 ----

    AsyncFunction("addLocalNotification") { (params: [String: Any]) in
      let notifId = (params["id"] as? NSNumber)?.stringValue ?? "0"
      let title = (params["title"] as? String) ?? ""
      let content = (params["content"] as? String) ?? ""
      let fireTime = (params["fireTime"] as? NSNumber)?.doubleValue ?? 0
      let extras = params["extras"] as? [String: Any]
      let category = params["category"] as? String
      ExpoJpushBridge.shared.addLocalNotification(
        id: notifId, title: title, content: content,
        fireTime: fireTime, extras: extras, category: category
      )
    }

    AsyncFunction("removeLocalNotification") { (params: [String: Any]) in
      guard let notifId = params["id"] as? String else { return }
      ExpoJpushBridge.shared.removeLocalNotification(id: notifId)
    }

    AsyncFunction("clearLocalNotifications") { () in
      ExpoJpushBridge.shared.clearLocalNotifications()
    }
  }
}
